### PR TITLE
implement support for __rust_alloc_error_handler

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/miri/issues/3439

### CODE REVIEW

File: tests/pass/alloc-access-tracking.stderr
Comment (saethlin) [inline tests/pass/alloc-access-tracking.stderr:1]: Do we care to change the tracked ID to keep this stderr file intact?

Comment (RalfJung) [reply tests/pass/alloc-access-tracking.stderr:1]: Ah, oops

File: tests/fail/alloc/alloc_error_handler.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//@normalize-stderr-test: "\| +\^+" -> "| ^"
+#![feature(allocator_api)]
+
+use std::alloc::*;
+use std::ptr::NonNull;
+
+struct BadAlloc;
+
+// Create a failing allocator; Miri's native allocator never fails so this is the only way to
+// actually call the alloc error handler.
```
Comment (bjorn3) [inline tests/fail/alloc/alloc_error_handler.rs:12]: User code can directly call `std::alloc::handle_alloc_error(layout)` on stable to call the alloc error handler.

**Pull request merged**